resources/type-system/flow-analysis.md

```diff
@@ -270,95 +262,105 @@ We also make use of the following auxiliary functions:
-       block is already modelled here).
-
-- `restrict(MB, MF, N)`, where `MB` and `MF` are flow models and `N` is a set of
-  variables assigned in the finally clause, models the flow of information
-  through a try/finally statement, and is defined as follows:
-
-  - We define `restrict(MB, MF, N)` to be `M3 = FlowModel(r3, VI3)` where:
-    - `MB = FlowModel(rb, VIB)`
-    - `MF = FlowModel(rf, VIF))`
-    - `pop(rb) = pop(rf) = r0` for some `r0`
-    - `r3` is `push(r0, top(rb) && top(rf))`
-    - `b` is true if `v` is in `N` and otherwise false
-    - `VI3` is the map which maps each variable `v` in the domain of `VIB` and
-      `VIF` to `restrictV(VIB(v), VIF(v), b)`.
+  For brevity, we will sometimes extend `join` to more than two arguments in the
+  obvious way.  For example, `join(M1, M2, M3)` represents `join(join(M1, M2),
+  M3)`, and `join(S)`, where S is a set of models, denotes the result of folding
+  all models in S together using `join`.
+
+  Simiarly, if L is a non-empty list of flow models, `join(...L)` represents the
+  result of folding the list using `join`.
+
+- `rebasePromotedTypes(p1, p2)`, where `p1` and `p2` are promotion chains,
+  computes a promotion chain containing promotions from both `p1` and `p2`,
+  discarding promotions if necessary to avoid violating the promotion chain
+  requirement. It is defined as `p3`, where:
+  - If `p2` is empty, then `p3 = p1`.
+  - Otherwise:
+    - Let `T2` be the last type in `p2`.
+    - Let `p1'` be a promotion chain obtained by deleting any elements `T` from
+      `p1` that do not satisfy `T <: T2`.
+    - Let `p3 = [...p2, ...p1']`.
+
+- `attachFinallyV(VM1, VM2, VM3)`, where `VM1`, `VM2`, and `VM3` are variable
+  models, represents the state of a variable model after a `try/finally`
+  statement, where `VM1` is the state after the `try` block, `VM2` is the state
+  before the `finally` block, and `VM3` is the state after the `finally`
+  block. It is defined as `VariableModel(d4, p4, s4, a4, u4, c4)`, where:
+  - Let `VM1 = VariableModel(d1, p1, s1, a1, u1, c1)`.
+  - Let `VM2 = VariableModel(d2, p2, s2, a2, u2, c2)`.
+  - Let `VM3 = VariableModel(d3, p3, s3, a3, u3, c3)`.
+  - Let `d4 = d3`. _A variable's declared type cannot change. Therefore, `d1 =
+    d2 = d3`, so it is safe to simply pick `d3`._
+  - Let `p4` be determined as follows:
+    - If the variable's value might have been changed by the `finally` block
+      (_TODO(paulberry): specify precisely how this is determined_), then `p4 =
+      p3`. _Promotions from the `try` block aren't necessarily valid, so only
+      promotions from the `finally` block are kept._
+    - Otherwise, `p4 = rebasePromotedTypes(p3, p1)`.
+  - Let `t4 = t3`. _The `finally` block inherited all tests from the `try`
+    block, so `t3` contains all relevant tested types._
+  - Let `a4 = a1 || a3`. _The variable is definitely assigned if it was
+    definitely assigned in either the `try` or the `finally` block._
```

Comment:
Is that 'definitely assigned at the end of the `try` block'? (So we're using the fact that the `try` block did not throw.)

Reply:
Correct.